design_notes/v2.md

```diff
@@ -0,0 +1,98 @@
+- [Re-implement loadable_variables using ManifestStore ](https://github.com/zarr-developers/VirtualiZarr/issues/473)
+- [Should we try to use Obstore for reading files?](https://github.com/zarr-developers/VirtualiZarr/issues/476)
+- [Challenges with earthaccess.open_virtual_mfdataset for grouped netCDFs with nested structure](https://github.com/zarr-developers/VirtualiZarr/issues/487)
+- [Docs on how to write a custom reader](https://github.com/zarr-developers/VirtualiZarr/issues/452)
+- [Make readers pluggable via entrypoint system](https://github.com/zarr-developers/VirtualiZarr/issues/245)
+- [Architecture diagram](https://github.com/zarr-developers/VirtualiZarr/issues/225)
+
+## Data structures
+
+- `ChunkManifest`: An object defining the paths, chunk byte offsets, lengths of one virtual array.
+- `ManifestArray`: A chunk manifest and its associated Zarr V3 ArrayMetadata
+- `ManifestGroup`: ManifestGroups are a way to organizing child groups and arrays into a data hierarchy. Contains Zarr `GroupMetadata` and one or more ManifestGroups and/or ManifestArrays.
+- `ManifestStore`: A ManifestGroup and its associated ObjectReader (see protocol below).
+- `xarray.Dataset`: Contains ManifestArray-backed variables and (optionally) in-memory NumPy-backed variables.
+
+## Definitions
+
+- `ObjectReader` - An implementation of the `GetAsync` and `GetRangeAsync` protocols defined in obspec. `ObjectReader`s are used to load arrays using `ManifestStore`.
+- `ReadableFile` - A reader that implements the `Read`, `Seek`, and `Tell` methods using a similar interface as a [Python BufferedReader](https://docs.python.org/3/library/io.html#io.BufferedReader). `ReadableFile`s are used by a `Backend` to construct a `ManifestStore`.
+- `Backend` - A backend is a function that constructs a `ManifestStore` from a `ReadableFile`, `ObjectReader`, and optionally `kwargs` specified using `virtual_backend_kwargs`.
+
+
+## Protocols
+
+```python
+from obspec import GetAsync, GetRangeAsync, Read, Seek, Tell
+
+class ObjectReader(GetAsync, GetRangeAsync):
+    ...
+
+class ReadableFile(Read, Seek, Tell):
+    ...
+```
+
+```python
+class Backend(protocol):
+    """
+    Callback protocol for creating a virtual zarr store from a file
+    """
+    def __call__(
+        file: ReadableFile,
+        object_reader: ObjectReader,
+        *,
+        virtual_backend_kwargs: dict | None = None,
+    ) → ManifestStore:
+        ...
+```
+
+## Top-level API functions
+
+- `open_virtual_dataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from a file
+- `open_virtual_mfdataset` - construct a Xarray Dataset with NumPy and/or ManifestArray-backed variables from many files
+
+> **NOTE**: Here the construction of a virtual dataset from a manifest store would happen in the top-level function rather than the backends.
+
+```python
+ def open_virtual_dataset(
+    readable_file: ReadableFile,
```

Comment:
@maxrjones I just started playing with a PR following this proposed design and the discussion from https://github.com/zarr-developers/VirtualiZarr/issues/553.  I think one additional issue is that our `ChunkManifest` generation requires the full path string to the file being virtualized.  Looking through the `obstore` code I don't think we a way to reconstruct this full path in a generic way from either a `ReadableFile` or an `ObjectReader` so we may also need to pass the explicit path string in addition (even though it seems duplicative 😆).  I'll way for @kylebarron to confirm this is the case.

Reply:
By "full path" you mean not just the path inside the bucket but also the protocol, bucket, and prefix? So you need `s3://bucket/path` or `abfs://container/path`?

The `ReadableFile` has a way to get the path not including the bucket and store prefix. https://developmentseed.org/obstore/latest/api/file/#obstore.ReadableFile.meta

And then the store allows you to get the `prefix` that was originally provided.

So you'd just need to keep a reference onto the store and infer the prefix yourself.

Reply:
> By "full path" you mean not just the path inside the bucket but also the protocol, bucket, and prefix?

Yes - the `ManifestArray` needs to hold all the path information that Icechunk will eventually need to read back from that object.

This is important - should the reader function signature actually be something more like:

```python
def reader(bucket_config: StoreConfig, *files: ReadableFile) -> ManifestStore:
    ...
```

Reply:
Since both ffspec and obstore have ways to create a readable file object from a store and a file url, I think this could work:
```
    def __call__(
        file_url: str,
        object_reader: ObjectReader,
        *,
        virtual_backend_kwargs: dict | None = None,
    ) → ManifestStore:

> This is important - should the reader function signature actually be something more like:

This is pretty close to a previous edition - it's was challenging to support interoperability between different ObjectReaders.